File: common-domain/src/main/java/com/netmarble/chat/domain/repository/ChatRoomMemberRepository.java
package com.netmarble.chat.domain.repository;

import com.netmarble.chat.domain.model.ChatRoomMember;

import java.util.Optional;
import java.util.Set;

/**
 * ChatRoomMember 도메인 리포지토리 인터페이스
 */
public interface ChatRoomMemberRepository {

    ChatRoomMember save(ChatRoomMember member);

    /**
     * 특정 채팅방의 활성 멤버 단건 조회 (lastReadMessage JOIN FETCH 포함)
     */
    Optional<ChatRoomMember> findActiveByChatRoomIdAndUserId(Long chatRoomId, Long userId);

    /**
     * 유저가 활성 멤버로 참가 중인 채팅방 ID 목록 조회 (단일 쿼리)
     */
    Set<Long> findActiveChatRoomIdsByUserId(Long userId);
}

File: server/chat-server/src/domain/repository/chat-room-member.repository.ts
import { ChatRoomMember } from '../model/chat-room-member.js';

export const CHAT_ROOM_MEMBER_REPOSITORY = Symbol(
  'ChatRoomMemberRepository',
);

export abstract class ChatRoomMemberRepository {
  abstract save(member: ChatRoomMember): Promise<ChatRoomMember>;
  abstract findActiveByChatRoomIdAndUserId(
    chatRoomId: number,
    userId: number,
  ): Promise<ChatRoomMember | null>;
  abstract findByChatRoomIdAndUserId(
    chatRoomId: number,
    userId: number,
  ): Promise<ChatRoomMember | null>;
  abstract findActiveChatRoomIdsByUserId(userId: number): Promise<Set<number>>;
  abstract findActiveByUserId(userId: number): Promise<ChatRoomMember[]>;
  abstract findActiveByChatRoomId(chatRoomId: number): Promise<ChatRoomMember[]>;
}
